src/io-queue.c

```diff
@@ -0,0 +1,224 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <sys/eventfd.h>
+#include <unistd.h>
+
+#include "xcp-ng/async-io/io-queue.h"
+#include "xcp-ng/async-io/io-req.h"
+
+// =============================================================================
+
+// Move [HEAD1, LAST] to the new list HEAD2.
+// It's an optimization to avoid remove/insert.
+#define STAILQ_CUT(HEAD1, LAST, HEAD2, FIELD) \
+  do { \
+    assert((LAST)); \
+    (HEAD2)->stqh_first = (HEAD1)->stqh_first; \
+    (HEAD2)->stqh_last = &(LAST)->FIELD.stqe_next; \
+    (LAST)->FIELD.stqe_next = NULL; \
+    if (!((HEAD1)->stqh_first = (LAST)->FIELD.stqe_next)) \
+      (HEAD1)->stqh_last = &(HEAD1)->stqh_first; \
+  } while (false)
+
+// -----------------------------------------------------------------------------
+
+// Call the request callback after completion.
+static inline void process_response (XcpIoReq *req, int res) {
+  int err;
+  if (XCP_UNLIKELY(res < 0))
+    err = res;
+  else if (XCP_LIKELY((size_t)res == xcp_io_req_get_size(req)))
+    err = 0;
+  else
+    err = -EIO;
+
+  req->cb(req, err, req->userData);
+}
+
+// Fetch responses in the queue and notify user.
+static inline unsigned int fetch_responses (XcpIoQueue *queue) {
+  struct io_uring *ring = &queue->pImpl.ring;
+
+  // How many responses are ready?
+  const unsigned int count = io_uring_cq_ready(ring);
+  if (XCP_UNLIKELY(!count))
+    return 0;
+  assert(count <= queue->capacity);
+
+  // Fill response array.
+  unsigned int head = *ring->cq.khead;
+  const unsigned int mask = *ring->cq.kring_mask;
+
+  for (const unsigned int last = head + count; head != last; ++head) {
+    const struct io_uring_cqe *cqe = &ring->cq.cqes[head & mask];
+    process_response((XcpIoReq *)cqe->user_data, cqe->res);
+  }
+
+  // Mark responses as read in the ring.
+  const struct io_uring_cq *cq = &ring->cq;
+  io_uring_smp_store_release(cq->khead, *cq->khead + count);
+
+  assert(queue->inflightCount >= count);
+  queue->inflightCount -= count;
+
+  return count;
+}
+
+// Cancel all given requests.
+static inline void cancel_requests (XcpIoReq *reqs, int err) {
+  while (reqs) {
+    XcpIoReq *nextReq = STAILQ_NEXT(reqs, pImpl.next);
+    reqs->cb(reqs, err, reqs->userData);
+    reqs = nextReq;
+  }
+}
+
+// Fill a io_uring_sqe instance from a XcpIoReq.
+static inline void set_sqe_from_req (const XcpIoReq *req, struct io_uring_sqe *sqe) {
+  sqe->opcode = req->opcode;
+  sqe->flags = 0;
+  sqe->ioprio = 0;
+  sqe->fd = req->fd;
+  sqe->off = (uint64_t)req->offset;
+  sqe->addr = (uint64_t)req->addr;
+  sqe->len = req->len;
+  sqe->rw_flags = 0;
+  sqe->user_data = (uint64_t)req;
```

Comment:
Why give the whole `req` and not just its `userData`?

Reply:
Because we must have a link to the original request in the cqe.
See: https://github.com/xcp-ng/xcp-ng-async-io/blob/feat/xcp-io-queue-impl/src/io-queue.c#L70
And: https://github.com/xcp-ng/xcp-ng-async-io/blob/feat/xcp-io-queue-impl/src/io-queue.c#L42